src/grt/grt-vpi.adb

```diff
@@ -176,51 +176,66 @@ package body Grt.Vpi is
       end case;
    end Trace_Cb_Reason;
 
-   procedure Trace_Property (V : Integer) is
+   -- Returns a string representation of a vpiType property
+   function Vpi_Type_To_String(V: Integer) return access constant String
+   is
    begin
       case V is
          when vpiUndefined =>
-            Trace ("vpiUndefined");
+            return new String'("vpiUndefined");
          when vpiType =>
-            Trace ("vpiType");
+            return new String'("vpiType");
          when vpiName =>
-            Trace ("vpiName");
+            return new String'("vpiName");
          when vpiFullName =>
-            Trace ("vpiFullName");
+            return new String'("vpiFullName");
          when vpiSize =>
-            Trace ("vpiSize");
+            return new String'("vpiSize");
          when vpiTimePrecision =>
-            Trace ("vpiTimePrecision");
+            return new String'("vpiTimePrecision");
          when vpiScalar =>
-            Trace ("vpiScalar");
+            return new String'("vpiScalar");
          when vpiVector =>
-            Trace ("vpiVector");
-
+            return new String'("vpiVector");
          when vpiModule =>
-            Trace ("vpiModule");
+            return new String'("vpiModule");
+         when vpiDefFile =>
+            return new String'("vpiDefFile");
          when vpiNet =>
-            Trace ("vpiNet");
+            return new String'("vpiNet");
+         when vpiPort =>
+            return new String'("vpiPort");
+         when vpiDirection =>
+            return new String'("vpiDirection");
          when vpiParameter =>
-            Trace ("vpiParameter");
+            return new String'("vpiParameter");
          when vpiScope =>
-            Trace ("vpiScope");
+            return new String'("vpiScope");
          when vpiInternalScope =>
-            Trace ("vpiInternalScope");
+            return new String'("vpiInternalScope");
          when vpiLeftRange =>
-            Trace ("vpiLeftRange");
+            return new String'("vpiLeftRange");
          when vpiRightRange =>
-            Trace ("vpiRightRange");
-
+            return new String'("vpiRightRange");
          when vpiStop =>
-            Trace ("vpiStop");
+            return new String'("vpiStop");
          when vpiFinish =>
-            Trace ("vpiFinish");
+            return new String'("vpiFinish");
          when vpiReset =>
-            Trace ("vpiReset");
-
+            return new String'("vpiReset");
          when others =>
-            Trace (V);
+            return null;
       end case;
+   end Vpi_Type_To_String;
+
+   procedure Trace_Property (V : Integer) is
+      Str: constant access constant String := Vpi_Type_To_String(V);
+   begin
+      if Str = null then
+         Trace(V);
+      else
+         Trace(Str.all);
+      end if;
    end Trace_Property;
```

Comment:
I know this is not definitive work, but why do you need Vpi_Type_To_String here ?  Trace_Property creates a memory leak.

Reply:
I was trying to be DRY here. The string representations are required both for vpi_get_str and Trace. What would be the correct way to get around this?

Reply:
In full Ada, you could simply return the string, but this is not supported in grt.

Simply allocate the strings only once:
```
vpiReset_Name : constant String_Cst := new String'("vpiReset");

...
  return VpiReset_Name;
...
```

I haven't tested that code (I will do it now), but IIRC the string will be pre-allocated.
I know, this is a little bit heavy.

Reply:
This compiles fine.

Reply:
Shouldn't the allocated memory be free once the last reference to it goes out of scope? I would greatly appreciate your help on this as Ada is something I still can't quite get my head around!
Travis is failing and I don't have a clue what change here is causing the VITAL test to fail.

Reply:
No, there is no garbage collection.  In Ada, memory allocated *could* be released when the access type declaration goes out of scope, but this is not the case here.

There is nothing fancy about Ada, in particular in grt.  This could be considered like C code, but with a different syntax.

IIRC, VITAL doesn't use vpi.  I will have a look at the Travis failure.

Reply:
I don't understand your change in grt-avhpi.adb: no need to add NUL here, this is done in grt-vpi.adb just after the call to vhpi_get_str.

Reply:
I was aware of the NUL termination being added in grt-vpi, but I thought maybe it would make further developments less prone to errors if the NUL is added at the source. I also don't think this affect correctness, and it's highly unlikely for it to cause any noticeable performance hit. But we can definitely remove it if you think it's not a good idea.

Reply:
Strangely the failing Travis test (gna/ticket30 on macOS) runs perfectly fine for me on macOS/llvm backend:

```
testsuite/gna/ticket30$ ./testsuite.sh 
analyze lib_numeric_tb.vhd
elaborate numeric_tb
simulate numeric_tb --stop-time=10ns --wave=numeric_tb.ghw --sdf=typ==lib_numeric_tb.sdf ( numeric_tb --stop-time=10ns --wave=numeric_tb.ghw --sdf=typ==lib_numeric_tb.sdf )
tperiod_CLK_posedge = -0.045 ns
Remove work library
Test successful

testsuite/gna/ticket30$ clang --version
Apple LLVM version 10.0.1 (clang-1001.0.37.7)
Target: x86_64-apple-darwin18.2.0

testsuite/gna/ticket30$ ghdl --version
GHDL 0.36-dev (20181129-183-g003ffb7c-dirty) [Dunoon edition]
 Compiled with GNAT Version: Community 2018 (20180523-73)
 llvm code generator
Written by Tristan Gingold.
```